Allow a non-OpenMP based build

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#19749 Allow a non-OpenMP based build**

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

`./build/bin/parallel_info`
`python -c "import torch; print(torch.__config__.parallel_info())"`

Differential Revision: [D15141993](https://our.internmc.facebook.com/intern/diff/D15141993)

Commit:
Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

dzhulgakov commented on binaries/parallel_info.cc:
```diff
@@ -0,0 +1,86 @@
+ * limitations under the License.
+ */
+
+#include "ATen/Parallel.h"
```
Ed recently expanded torch.__config__.show() in https://github.com/pytorch/pytorch/pull/18579

It'd be nice to expose it to python as well for debugging, but I wonder what's the good name is (or just fold it into __config__.show)

ilia-cher replied:
nice, I'll check it and might merge some functionality with ATen/Version.h

ilia-cher replied:
I think config.show is more for build config and should be constant, but information about parallelism can change, e.g. depends on environment vars, or hardware.

I could add directly into `torch/__init__.py` through e.g. `torch._C.parallel_info()`

ilia-cher replied:
and implement parallel_info() in ATen/Parallel

dzhulgakov commented on cmake/Modules/FindMKLDNN.cmake:
```diff
@@ -89,7 +89,7 @@ ENDIF(MKL_FOUND)
 IF(MKL_FOUND)
   SET(MKL_cmake_included TRUE)
-  SET(MKLDNN_THREADING "OMP:COMP" CACHE STRING "" FORCE)
+  SET(MKLDNN_THREADING "OMP:COMP" CACHE STRING "")
```
what is this change? (asking as cmake noob)

ilia-cher replied:
without FORCE it prefers an already set value of MKLDNN_THREADING, with FORCE it overwrites it. I'm a cmake noob myself, so if @kostmo could double check that'd be great

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info
python -c "import torch; print(torch.parallel_info())"

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

ilia-cher commented:
Now we can also print parallel info with
```
python -c "import torch; print(torch.parallel_info())"
```
merged some code with ATen/Version

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info
python -c "import torch; print(torch.parallel_info())"

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info
python -c "import torch; print(torch.parallel_info())"

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info
python -c "import torch; print(torch.parallel_info())"

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info
python -c "import torch; print(torch.parallel_info())"

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info
python -c "import torch; print(torch.parallel_info())"

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info
python -c "import torch; print(torch.parallel_info())"

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info
python -c "import torch; print(torch.parallel_info())"

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

dzhulgakov commented on torch/__init__.py:
```diff
@@ -309,6 +309,10 @@ def compiled_with_cxx11_abi():
     r"""Returns whether PyTorch was built with _GLIBCXX_USE_CXX11_ABI=1"""
     return _C._GLIBCXX_USE_CXX11_ABI
 
+def parallel_info():
```
the only style question I have is `torch.parallel_info()` vs `torch.__config__.parallel_info()`. @gchanan  - thoughts?

fmassa replied:
I'd vote for not adding it in the torch namespace,  but instead in `__config__`.

gchanan replied:
I'd put it in `__config__`.

dzhulgakov reviewed (approved):
Otherwise looks good assuming tests pass and the output looks legible

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info
python -c "import torch; print(torch.parallel_info())"

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info
python -c "import torch; print(torch.parallel_info())"

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info
python -c "import torch; print(torch.parallel_info())"

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info
python -c "import torch; print(torch.parallel_info())"

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

ilia-cher commented:
Moved to `torch.__config__`

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info
python -c "import torch; print(torch.parallel_info())"

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info
python -c "import torch; print(torch.parallel_info())"

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head

Commit:
Update on "Allow a non-OpenMP based build"

Allow a non-OpenMP based build

Summary:
A few changes:
 - fix in torch/CMakeLists.txt to honor USE_OPENMP
 - adding MKL_SEQ=1 to choose sequential MKL
 - forwarding MKLDNN_THREADING to control threading in MKL-DNN
Also added a small diagnostic utility that prints parallellization info

Test Plan:
BUILD_BINARY=1 MKLDNN_THREADING=SEQ USE_OPENMP=0 MKL_SEQ=1  BLAS=MKL
USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake

./build/bin/parallel_info
python -c "import torch; print(torch.parallel_info())"

gh-metadata: pytorch pytorch 19749 gh/ilia-cher/5/head
